docs/Migration-Guide-V3.md

```diff
@@ -0,0 +1,186 @@
+# V3 Migration Guide
+
+This guide is aimed to help migration from Fastify v2 to v3.
+
+Before beginning please ensure that any deprecation warnings from v2 are fixed as we have removed all deprecated features and they will no longer work after upgrading. ([#1750](https://github.com/fastify/fastify/pull/1750))
+
+## Breaking changes
+
+### Changed middleware support ([#2014](https://github.com/fastify/fastify/pull/2014))
+
+From Fastify v3, middleware support does not come out of the box with the framework itself.
+
+If you use Express middleware in your application, please install and register the [`fastify-express`](https://github.com/fastify/fastify-express) or [`middie`](https://github.com/fastify/middie) plugin before doing so.
+
+**v2:**
+
+```js
+fastify.use(require('cors')());
+```
+
+**v3:**
+
+```js
+await fastify.register(require('fastify-express'));
+fastify.use(require('cors')());
+```
+
+### Changed logging serialization ([#2017](https://github.com/fastify/fastify/pull/2017))
+
+We have updated our logging [Serializers](https://github.com/fastify/fastify/blob/master/docs/Logging.md) to now receive Fastify [`Request`](https://github.com/fastify/fastify/blob/master/docs/Request.md) and [`Reply`](https://github.com/fastify/fastify/blob/master/docs/Reply.md) objects instead of native ones.
+
+If you have created custom serializers they will need updating if they expect properties that aren't exposed by the Fastify objects themselves.
+
+**v2:**
+
+```js
+const fastify = require('fastify')({
+  logger: {
+    serializers: {
+      res(res) {
+        return {
+          statusCode: res.statusCode,
+          customProp: res.customProp
+        };
+      }
+    }
+  }
+});
+```
+
+**v3:**
+
+```js
+const fastify = require('fastify')({
+  logger: {
+    serializers: {
+      res(reply) {
+        return {
+          statusCode: reply.statusCode, // No change required
+          customProp: reply.raw.customProp // Log custom property from res object
+        };
+      }
+    }
+  }
+});
+```
+
+### Changed schema substitution ([#2023](https://github.com/fastify/fastify/pull/2023))
```

Comment:
We also used to support `@hapi/joi`. I think the schema-replacement functionality was helping Joi to reuse schemas. As a result, we do not support this anymore. Can you investigate and add a paragraph about it?

Reply:
This is still working but I've fixed the code snippet in the docs